hashi_vault refresh - Add AWS login methods, bugfixes, cleanup

briantist opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
### **This is a re-submission of https://github.com/ansible/ansible/pull/66735**
<s>The original had a changelog fragment but I don't see where to put that in the new repo.</s> 
New changelog added.

----
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

This PR adds support for the AWS IAM Login auth method in Hashicorp Vault. Along the way ran into some small bugs and possible improvements and so it ended up being a larger change.

## Outward Facing Changes
- Added INI sources for some of the parameters like the vault URL, role ID, and some of the newer params.
- [BUG] Previously, this plugin didn't work correctly when used with `with_`. If you passed `with_hashi_vault` a list of terms strings, it would look up the first one, and then for each subsequent one it would ignore what was passed and return the first result again. This is fixed.
- When passing params through the term string, you can now pass secrets as an unnamed argument, if it comes first. Example: ``{{ lookup('hashi_vault', 'secret=secret/hello:value') }}`` is the same as `{{ lookup('hashi_vault', 'secret/hello:value') }}`. Nicer syntax overall, and works well with `with_` if you have a large list of secrets for some reason.
- New option `return_format` with alias `as` can be used to control how a secret listing is returned (especially when you ask for a path that contains multiple key/value returns). Previous behavior was to return the single `dict` of key-value pairs. This is now the default. The choices are:
    * `dict` (same as previous behavior)
    * `raw`: return the raw output from the API, including metadata
    * `values`: return a list of all values (discard keys)
  
  `raw` is useful if you need the metadata from a secret. `values` is friendlier in loops when keys are unimportant.
- All parameters can now be passed in as params, not just in the term string, so you can do: `{{ lookup('hashi_vault', 'secret/hello:value', auth_method='ldap', username='itsame', password=secret_var) }}`
- While token auth was available before, it was not listed explicitly in the `auth_method` option choices, so now it is the "official" default (it already was the default before effectively, so operation won't change).
- If vault token was not supplied before, it was looked up in `~/.vault-token` (which is usually set by the vault CLI), however this wasn't documented. This is now explicit, and there are 2 options which you can set to control where it looks for the file, `token_path` (the directory) and `token_file`.
- `token_path` also retrieves its value from `$HOME` effectively making that the default, which also preserves previous behavior, and the value can be set in INI as well.
- `token_file` defaults to `.vault-token` and can also be set in INI.
- If your version of the HVAC library uses a deprecated method name for your chosen `auth_type`,  you'll see deprecation warnings because they changed the interface in newer versions. The plugin now checks and calls the newest one available.
- [BUG] sometimes trying to use a specific auth method with invalid params / credentials would implicitly use a token, even when not supplied, by loading it from env or file. This is no longer the case. `token` is only used now when `auth_method == 'token'` either by explicitly setting it or by not setting any `auth_method` (default is `token`).
- AWS IAM auth method is added. ~This plugin now pulls in the standard aws auth options via doc fragments.~ `<--` since that would cause a cross-collection dependency (if it's even possible) I've just copied the relevant options for now.
- If you using IAM auth without explicit credentials, botocore is used to try to load credentials. If trying to use a boto profile, boto3 is used. If neither of those methods are used, the boto libraries are not required and not loaded.
- Documentation and examples are updated
- Friendly error messages restored / added for non-existent secrets and permission denied errors (fixes #54)

## Internal Changes
- Improved term processing to account for term array, for secret being unnamed if first
- All options loaded through the `LookupBase` provided options methods. This is a huge improvement that eliminates all manual checking for env vars, required params, ini values, aliases, etc., as the LookupBase helper method does all this for us, based on the documentation of options.
- Some refactoring, putting initial term and options processing into methods, and adding individual validation methods for each auth method that check required options and do deep argument transformation where necessary (loading `token` from file for example).
- Creating the `HvacClient` does not automatically authenticate now; a separate `.authenticate()` method has been added. This is paving the way for re-use of an authenticated connection to retrieve several/many secrets, though that pattern won't be implemented in this PR.
- `get()` method always returns a list now, even for single items. This makes it easier to provide a flattened list from the plugin by extending return value list.
- removed no longer used `**kwargs` from authentication methods.
- new internal comments on how to add authentication methods.
- new tests added for return format that also use options outside the term string, and secret paths without `secret=`.
- a kv v2 path containing multiple k/v pairs has been added in the tests which can be used with more tests from other contributors.
- [FUTURE] Also thinking about moving the HVAC class to a module_util so that it can be shared among other plugins and modules later, but this won't be in this PR.


##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Feature Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
hashi_vault.py

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->

## Compatibility
Despite the changes, as far as I can tell the plugin will remain 100%* backward compatible with the current iteration (**NB:** based on the state of devel/master before this PR, which does change behavior from 2.9 after https://github.com/ansible/ansible/pull/64288 was merged), so all existing uses and calls should still work *(with the exception of using `with_` with more than 1 value, because it was returning incorrect data before).

gundalow commented:
Hi @ihard @tomcooperca @flowerysong I believe you've helped on the `hashi_vault` inventory plugin before. 

@briantist has raised this PR which adds support for the AWS IAM Login auth method in Hashicorp Vault. Along the way ran into some small bugs and possible improvements and so it ended up being a larger change.

I was wondering if you'd be interested in taking a look and helping it progress?

Thanks in advance.

briantist commented:
Thanks @gundalow !

It's my understanding that the tests here in community.general are still being worked on. Just want to mention the tests for these changes were passing in the old repo, but I understand I may have to update them for things like fully qualifying names, etc.

Ready and able to make whatever changes are necessary.

Also you mentioned an inventory plugin for `hashi_vault`? Not sure if that was a mistake or not but I am very interested in an inventory plugin for that purpose so I'd appreciate a link to it if there is one. 

Some of the internal changes I made here which are working toward moving some of the hashi vault code to be reusable (possibly in a module util) might lend itself well to being shared with an inventory plugin.

gundalow commented:
> It's my understanding that the tests here in community.general are still being worked on ( #29 maybe others). Just want to mention the tests for these changes were passing in the old repo, but I understand I may have to update them for things like fully qualifying names, etc.

@briantist Tests for community.general are unstable at the moment, so thank you for your understanding at the moment.

 
> Ready and able to make whatever changes are necessary.
> 
> Also you mentioned an inventory plugin for `hashi_vault`? Not sure if that was a mistake or not but I am very interested in an inventory plugin for that purpose so I'd appreciate a link to it if there is one.

@briantist Sorry, you are correct. I meant `lookup plugin`.

briantist commented:
kicking the tests since #29 landed, even if they're not 100% yet would like to see if it surfaces anything that I specifically need to fix for this PR.

felixfontein commented:
Thanks for doing that! I'm very curious to see how the tests look now for a "regular" PR.

briantist commented:
> Thanks for doing that! I'm very curious to see how the tests look now for a "regular" PR.
@felixfontein it looks like most of the failures were for other stuff unrelated to this, but the sanity checks did highlight the AWS doc fragments I was trying to use. I've just removed those and duplicated the params. Also updated to remove `version_added`.

briantist commented:
Test kick for #42

gundalow commented:
@onitake Hi, would you possibly be able to review this PR? In particular the `plugins/lookup/hashi_vault.py` file.

If you are also able to copy this file locally and test it that would give us confidence to be able to merge it.

Thanks in advance for any help!

briantist commented:
> @onitake Hi, would you possibly be able to review this PR? In particular the `plugins/lookup/hashi_vault.py` file.

@onitake I can lend a hand with testing locally if needed. An easy way to do that is copy the `hashi_vault.py` file into the `library/` directory alongside your playbook. 

[Here's the direct link to the file](https://github.com/ansible-collections/community.general/blob/58953e2d60d437d7cbab875a8f146ca5c916077d/plugins/lookup/hashi_vault.py), and you click **Raw** to copy/paste, or right click **Raw** to download the file directly.

I like to rename the file to something else so you get a "new" plugin name, like `test_hashi_vault.py`. That will let you call both `lookup('test_hashi_vault', ...)` (version from the PR) and `lookup('hashi_vault', ...)` (original) in the same playbook, and gives some confidence the new one is actually being loaded.

Then you can make whatever kind of playbook you want to test it, as long as the references are updated. Here's an example but of course you can update with your own params and secret names:

```
---
- hosts: localhost
  gather_facts: no
  tasks:
    - debug:
        msg: "{{ lookup('test_hash_vault', item) }}"
      loop:
        - secret/example:secret
        - secret/another
        - secret/i-cant-access
```

Don't hesitate to reach out if you need assistance or run into any issues.
> 
> Thanks in advance for any help!

felixfontein commented:
Please add a [changelog fragment](https://docs.ansible.com/ansible/latest/community/development_process.html#changelogs-how-to) in changelogs/fragments/.

(You probably have to rebase with current `master` to see that directory. You can also simply create it yourself without rebasing; git does not care about directories, only about files.)

onitake commented:
@briantist @gundalow Don't worry, I know how use custom modules in Ansible. :smiley: 

I'll give this a spin, but I can only test some aspects of the module.
For example, I can't test AWS authentication.

I should also note that my original request was only for better error reporting, not for new options, so I will mainly focus on that.

briantist commented:
> @briantist Don't worry, I know how use custom modules in Ansible. 😃

Heh sorry @onitake , didn't want to make any assumptions! 😅

> I'll give this a spin, but I can only test some aspects of the module.
> For example, I can't test AWS authentication.
> 
> I should also note that my original request was only for better error reporting, not for new options, so I will mainly focus on that.

You might also like the changes to how you can specify parameters and secrets, as they apply to everyone.

But anything is helpful thanks so much!

briantist commented:
> Please add a [changelog fragment](https://docs.ansible.com/ansible/latest/community/development_process.html#changelogs-how-to) in changelogs/fragments/.
> 
> (You probably have to rebase with current `master` to see that directory. You can also simply create it yourself without rebasing; git does not care about directories, only about files.)

✔changelog added; thanks for pointing out the dir location, @felixfontein . I was  looking for it when I first migrated since I had a chagelog before.

briantist commented:
woo! all tests passed except the "unstable" osx tests, thanks @felixfontein for all the work on tests for this collection!

briantist commented:
ready_for_review

felixfontein commented on changelogs/fragments/23-hashi-vault-lookup-refresh.yaml:
```diff
@@ -0,0 +1,13 @@
+bugfixes:
+  - hashi_vault - when a non-token authentication method like ldap or userpass failed, but a valid token was loaded anyway (via env or token file), the token was used to attempt authentication, hiding the failure of the requested auth method.
+  - hashi_vault - if used via `with_hashi_vault` and a list of (say) 4 secrets to retrieve, only the first one would be retrieved and returned 4 times.
```
```suggestion
  - hashi_vault - if used via ``with_hashi_vault`` and a list of (say) 4 secrets to retrieve, only the first one would be retrieved and returned 4 times.
```
This is ReST, not MarkDown. (Same for other entries.)

briantist replied:
Got it, changed all of them in the most recent commit. Thanks!

felixfontein commented on changelogs/fragments/23-hashi-vault-lookup-refresh.yaml:
```diff
@@ -0,0 +1,13 @@
+bugfixes:
+  - hashi_vault - when a non-token authentication method like ldap or userpass failed, but a valid token was loaded anyway (via env or token file), the token was used to attempt authentication, hiding the failure of the requested auth method.
+  - hashi_vault - if used via `with_hashi_vault` and a list of (say) 4 secrets to retrieve, only the first one would be retrieved and returned 4 times.
+  - hashi_vault - error messages are not user friendly and don't contain the secret name ( https://github.com/ansible-collections/community.general/issues/54 )
```
```suggestion
  - hashi_vault - error messages are not user friendly and don't contain the secret name ( https://github.com/ansible-collections/community.general/issues/54 )
```
Did your change make the error messages not user friendly? If you improved them, you should say so :)

briantist replied:
Should have been `now`, I don't know why I make this type so often, there are 2 keys between `w` and `t`! (fixed)

felixfontein commented on changelogs/fragments/23-hashi-vault-lookup-refresh.yaml:
```diff
@@ -0,0 +1,13 @@
+  - hashi_vault - `secret` can now be an unnamed argument if it's specified first in the term string (see examples).
+  - hashi_vault - previously all options had to be supplied via key=value pairs in the term string; now a mix of string and parameters can be specified (see examples).
+  - hashi_vault - new option `return_format` added to control how secrets are returned, including options for multiple secrets and returning raw values with metadata.
+  - hashi_vault - `token` is now an explicit option (and the default) in the choices for `auth_method`
```
What was the previous default value? Changing default values is always problematic, as it breaks backwards compatibility.

briantist replied:
I've updated the message to hopefully make this clearer, but I've gone to great lengths to preserve backwards compatibility for all of the changes in this PR.

In this case, the previous default was `None`, and the behavior from that was to use token auth (`token` was not in the `choices` before, you had to omit it).

So what I've changed is to add `token` as an explicit option in the choices, while also making it the default value, so this keeps previous _behavior_ and old playbooks will still work, but it now allows one to also use `auth_method: token` which previously would not have worked.

Let me know if the new changelog message is clear; if not I will not try to reword it.

felixfontein replied:
Sounds good to me!

felixfontein commented on plugins/lookup/hashi_vault.py:
```diff
@@ -1,109 +1,229 @@
   notes:
     - Due to a current limitation in the HVAC library there won't necessarily be an error if a bad endpoint is specified.
-    - As of Ansible 2.10, only the latest secret is returned when specifying a KV v2 path.
+    - As of Ansible 2.10, only the latest version of a secret is returned when specifying a KV v2 path.
```
@gundalow how should we write things like this? `As of community.general ...`? I'm not really sure...

briantist replied:
I saw some discussion about this a number of weeks ago (maybe on IRC or in one of the issues tracking these things). There was no consensus at that time, but one of the suggestions was that using Ansible 2.10 was likely ok as a marker of the first collection release.

I think it also has not been decided what the initial collection version numbers will be either.

If anything is decided or the guidance changes I'm happy to update this quickly.

felixfontein replied:
I guess we can keep it for now. At least `Ansible 2.10` can be searched+replaced if we decide to change it.

felixfontein commented on plugins/lookup/hashi_vault.py:
```diff
@@ -1,109 +1,229 @@
-      - Authentication method to be used.
-      - C(userpass) is added in version 2.8.
+        - Authentication method to be used.
+        - C(userpass) is added in version 2.8.
```
```suggestion
        - C(userpass) is added in Ansible 2.8.
```
To make it clear that you're not talking about the community.general version.

briantist replied:
Done!

briantist commented:
Thanks for the feedback @felixfontein ! I think I've addressed all your concerns and suggestions; let me know if there's anything I missed.

ihard commented:
As a suggestion - in my opinion, a very popular option would be to set the default value of the secret, for example, if there is no secret in the Vault. Using the lookup construct('hashi_vault', 'secret/not exist')|d (") does not work for lookup plugins as a rule.

briantist commented:
> As a suggestion - in my opinion, a very popular option would be to set the default value of the secret, for example, if there is no secret in the Vault. Using the lookup construct('hashi_vault', 'secret/not exist')|d (") does not work for lookup plugins as a rule.

@ihard That's a great idea! Yeah it usually wouldn't work with lookups since they generally return lists so won't trigger default.

I'm hesitant to add it to this PR only because it's already quite big, and I've been trying to get it merged for almost 3 months (of course the collections split is part of the delay).

Would you put in a feature request for it and tag me? It's something I'd be likely to take on after this lands.

felixfontein commented:
> Using the lookup `construct('hashi_vault', 'secret/not exist')|d (")` does not work for lookup plugins as a rule.

Have you tried `construct('hashi_vault', 'secret/not exist')|d (", true)`? (See https://jinja.palletsprojects.com/en/2.10.x/templates/#default)

felixfontein commented:
@ihard @mbrancato can you test this PR? At least whether it still works for your use-cases?

briantist commented:
> > Using the lookup `construct('hashi_vault', 'secret/not exist')|d (")` does not work for lookup plugins as a rule.
> 
> Have you tried `construct('hashi_vault', 'secret/not exist')|d (", true)`? (See https://jinja.palletsprojects.com/en/2.10.x/templates/#default)

That looks usable; now that I think of it, for this plugin at least a not found is an error, so you'd also need `errors='ignore'` or `errors='warn'`.

ihard commented:
I tested this PR on my playbook and came across a modified output of variables. 
ansible 2.9.4
I use it like this
```
    - name: set var_test
      set_fact:
        var_test: "{{ lookup('hashi_vault', 'secret=/secret_dev/data/my/test') }}"

    - debug:
        var: var_test
```
output old version

```
TASK [debug] *********************************************************************************************************************************************************************
ok: [localhost] => {
    "var_test": {
        "data": {
            "value": "123"
        },
        "metadata": {
            "created_time": "2020-04-08T16:25:21.159022899Z",
            "deletion_time": "",
            "destroyed": false,
            "version": 1
        }
    }
}
```

version from PR
```
TASK [debug] *********************************************************************************************************************************************************************
ok: [localhost] => {
    "var_test": {
        "value": "123"
    }
}
```

flowerysong commented on changelogs/fragments/23-hashi-vault-lookup-refresh.yaml:
```diff
@@ -0,0 +1,13 @@
+bugfixes:
```
These changelog entries are a mix of `bugfixes` and `minor_changes`, and should be separated out accordingly.

briantist replied:
@flowerysong 
imo all 3 of the things listed under `bugfixes` are in fact bugfixes
Which one(s) do you think should be moved to `minor_changes`?

felixfontein replied:
I think @flowerysong means that there should be two separate PRs, one for bugfixes, one for features.

flowerysong replied:
No, I somehow completely missed that you already had them split into two lists, and thought you had put everything in the `bugfixes` list.

felixfontein replied:
Oh ok, then it should be fine :)

flowerysong commented on changelogs/fragments/23-hashi-vault-lookup-refresh.yaml:
```diff
@@ -0,0 +1,13 @@
+bugfixes:
+  - hashi_vault - when a non-token authentication method like ldap or userpass failed, but a valid token was loaded anyway (via env or token file), the token was used to attempt authentication, hiding the failure of the requested auth method.
+  - hashi_vault - if used via ``with_hashi_vault`` and a list of (say) 4 secrets to retrieve, only the first one would be retrieved and returned 4 times.
```
```suggestion
  - hashi_vault - if used via ``with_hashi_vault`` and a list of n secrets to retrieve, only the first one would be retrieved and returned n times.
```

briantist commented:
> I tested this PR on my playbook and came across a modified output of variables.

Thanks @ihard this looks pretty important. I'm going to look into this and ensure there's consistency.

Could you tell me how `/secret_dev/data/my/test` is structured in vault? is it KV? v1 or v2? Anything else I could use to repro?

ihard commented:
In Vault 
secret_dev - this mount_point
my/test - path to secret
value - key
i use KV v2

felixfontein commented:
@ihard can you test your playbook against the `master` branch of this repository? I.e. whether this is actually caused by this PR, or by something before that? I remember that there was a change for v2 secrets ealier in the Ansible main repo. Maybe this is related to that.

felixfontein commented:
It's this one: ansible/ansible#64288

CC'ing @tpounds and @Im0 as they are (were?) also interested in this plugin.

briantist commented:
> It's this one: [ansible/ansible#64288](https://github.com/ansible/ansible/pull/64288)
> 
> CC'ing @tpounds and @Im0 as they are (were?) also interested in this plugin.

Ahh thank you so much for finding this. I thought this was the case but was failing to find the PR. 

Yes my change comes after that (that's where the output was changed), so mine follows the behavior of that one. In fact that PR is also what introduced the regression that led to #54 .

ihard commented:
I checked with the master branch of this repository - Yes, the output is similar. But I don't see these changes either in version 2.9.6 (current at the moment). These changes will be included in version 2.10?

felixfontein commented:
@ihard I think it has not been backported to stable-2.9, so the first version this will appear in is the version in this repository (i.e. in Ansible 2.10).

Im0 commented:
Thanks @felixfontein .  Unfortunately at the moment I'm not in a position to test this.  I'll keep an eye on this thread incase I get some time to test.

gundalow commented:
> @briantist @gundalow Don't worry, I know how to use custom modules in Ansible. 
> 
> I'll give this a spin, but I can only test some aspects of the module.
> For example, I can't test AWS authentication.
> 
> I should also note that my original request was only for better error reporting, not for new options, so I will mainly focus on that.

Hi @onitake, I was wondering if you have the chance to take a look at this PR?

onitake commented:
@gundalow As originally announced, I can only test the new output handling, but not the new authentication schemes.

This:
```
fatal: [localhost]: FAILED! => {"msg": "An unhandled exception occurred while running the lookup plugin 'hashi_vault'. Error was a <class 'ansible.errors.AnsibleError'>, original message: The secret '/secret' doesn't seem to exist."}
```
is much better than the old generic message. So, LGTM for me.

felixfontein commented:
rebuild_merge

felixfontein commented:
rebuild_merge

felixfontein commented:
@briantist thanks for doing this work!
@onitake @ihard @Im0 @flowerysong thanks for your feedback!

briantist commented:
Thanks for all your help through this @felixfontein , much appreciated!

flowerysong commented on tests/integration/targets/lookup_hashi_vault/lookup_hashi_vault/tasks/token_test.yml:
```diff
@@ -25,6 +28,32 @@
+    - name: "Check multiple secrets as dict"
+      fail:
+        msg: 'Return value was not dict or items do not match.'
+      when: (kv2_secrets_as_dict | type_debug != 'dict') or (kv2_secrets_as_dict['value{{ item }}'] != 'foo{{ item }}')
```
@briantist I missed this when I was skimming through this PR before, but noticed it when I was updating the tests. For future reference, things like this should be written using concatenation instead of relying on double templating. Conditionals like `when` are implicitly templated, so putting `{{ }}` in them 'works' but is rarely necessary and can end up doing confusing things.

```
when: (kv2_secrets_as_dict | type_debug != 'dict') or (kv2_secrets_as_dict['value' ~ item ] != 'foo' ~ item)
```

briantist replied:
Thanks @flowerysong I'm aware of that conditionals are already templated, though for the life of me I can't remember why I used templating there. Will keep it mind!

briantist commented:
Previously not called out, this PR also fixes #767
